### PR TITLE
[Metrics] Cache metrics ports in a file at each node

### LINKED ISSUE #13372: [Metrics] Metrics are not exported when using different jobs
### What is the problem?
In an ipython window, run
```python
import ray
from ray.util import metrics
ray.init(_metrics_export_port=9999)
```
and leave it running.  In a second terminal, in an ipython window run
```python 
import ray
from ray.util import metrics

ray.init(address="auto")
error_counter = metrics.Count( 
            "fake_error_counter", 
            description=("The number of exceptions that have " 
                         "occurred in the backend."), 
            tag_keys=("backend", )) 
error_counter.set_default_tags({"backend": "FAKE BACKEND"}) 

error_counter.record(12345)
```

If you check localhost:9999, this metric never gets printed.  However, if you now go back to the first ipython window and paste in a different custom metric (say `fake_error_counter_2`), that metric will successfully be printed.  

*Ray version and other system information (Python version, TensorFlow version, OS):*
Ray nightly, Python 3.6, Mac OS Big Sur
### Reproduction (REQUIRED)
Please provide a short code snippet (less than 50 lines if possible) that can be copy-pasted to reproduce the issue. The snippet should have **no external library dependencies** (i.e., use fake or mock data / environments):

If the code snippet cannot be run by itself, the issue will be closed with "needs-repro-script".

- [x] I have verified my script runs in a clean environment and reproduces the issue.
- [x] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

### PR DESCRIPTION
**Pull request by architkulkarni**
## Why are these changes needed?
Different drivers started on the same node need to use the same metrics port.  Before this PR, each subsequent driver script after the first would choose a new unused port for the metrics agent, so metrics defined in these driver scripts would not display.

## Related issue number

Closes https://github.com/ray-project/ray/issues/13372
## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/node.py
Comment (rkooo567) [inline python/ray/node.py:588]: This syntax is not available in python 3.6 I think?

File: python/ray/node.py
Comment (rkooo567) [inline python/ray/node.py:193]: We don't need this block right?

File: python/ray/node.py
Comment (architkulkarni) [reply python/ray/node.py:588]: My Dev environment is 3.6 so I think it's okay!

File: python/ray/node.py
Comment (architkulkarni) [reply python/ray/node.py:193]: Oh, didn't we say that whatever approach we do for metrics_agent_port we should do for metrics_export_port too? I'm not fully familiar with the second one yet so it's your call. Also, I'll do the rename of metrics_agent to dashboard_agent in a separate PR so that this diff is easier to read.

Comment (rkooo567) [reply python/ray/node.py:193]: You already assigned a port of this in line 190 right? Am I missing something?

File: python/ray/node.py
Comment (rkooo567) [reply python/ray/node.py:588]: Hmm not sure what was the issue before, but I definitely remember this wasn't supported... Let's see if CI says something

File: python/ray/node.py
Comment (rkooo567) [reply python/ray/node.py:193]: Line 193-195 will make line 190 meaningless IIUC?

Comment (architkulkarni) [reply python/ray/node.py:193]: Ah I see, you're totally right, I forgot to remove lines 193-195.  I'll do that

**Comment by architkulkarni** [comment]
`pytest test_metrics.py` fails on `test_metrics_export_end_to_end`, but `pytest test_metrics.py::test_metrics_export_end_to_end` succeeds... investigating

**Comment by architkulkarni** [comment]
Yeah, something is not working as expected in either `ray.shutdown()` or `cluster.shutdown()`, not sure exactly what yet.  On this branch the following code succeeds for `num_iterations = 1` but fails for `num_iterations = 2` on the second iteration:

```python
import ray
from ray.cluster_utils import Cluster
from ray.util.metrics import Count
from ray.test_utils import wait_for_condition, fetch_prometheus

num_iterations = 2
counters = [None] * num_iterations

for i in range(num_iterations):
    # Start a cluster, record and test a metric, and shut it down.

    print("STARTING UP")
    cluster = Cluster()
    cluster.add_node()
    cluster.wait_for_nodes()
    ray.init(address=cluster.address)

    counter_name = f"test_driver_counter_{i}"
    counters[i] = Count(counter_name, description="desc")
    counters[i].record(i)

    def test():
        node_info_list = ray.nodes()
        prom_addresses = []
        for node_info in node_info_list:
            metrics_export_port = node_info["MetricsExportPort"]
            addr = node_info["NodeManagerAddress"]
            prom_addresses.append(f"{addr}:{metrics_export_port}")

        components_dict, metric_names, metric_samples = fetch_prometheus(
                prom_addresses)
        return any(counter_name in full_name for full_name in metric_names)

    wait_for_condition(test, timeout=30)

    print("SHUTTING DOWN")
    ray.shutdown()
    cluster.shutdown()
```

**Comment by architkulkarni** [comment]
It seems the issue is just related to `ray.shutdown()`.  Removing all the `cluster_utils` stuff and just using `ray.init()` doesn't change anything; it still works with one iteration but not two.

**Review by edoakes — COMMENTED** [review]
I'm a little bit confused -- is this for the `address="auto"` case or for the "new driver" case? If it's for the `address="auto"` case the most logical solution would be to store the port mapping in the GCS and fetch it on startup. I'm not sure we want this behavior for the "new driver" case at all.

I think you mentioned there's some reason not to do that, could you remind me what that is?

**Comment by architkulkarni** [comment]
No, I think I might be confused--what's the difference between the two cases?  This PR is for when more than one driver script runs on the same Ray cluster and on the same node, so there's a new driver that calls `address="auto"`.  Maybe I used the term "job" incorrectly in the title.

Or maybe the source of confusion is that this PR currently fails tests because of a separate issue https://github.com/ray-project/ray/issues/13532 which involves the stats not shutting down correctly (in particular, in between different test functions when running pytest)

**Comment by edoakes** [comment]
The difference is that in the `address=auto` case, the metrics agent is started only once (when ray start is called) but in the `ray.init()` case it's started twice -- in the second scenario we probably don't want this caching behavior.

Assuming this is just solving the first case, the solution seems reasonable but ideally we would store this in the GCS instead of a local file for consistency.

**Comment by architkulkarni** [comment]
Ah, I think I see what you're saying.  The cache is keyed by session, so in the case of calling ray.init(), ray.shutdown(), and then ray.init(), the ports from the first ray.init() are not read by the second ray.init().  Does that address the concern?

**Review by rkooo567 — APPROVED** [review]

**Comment by rkooo567** [comment]
Btw, @allenyin55 said this should be fixed ASAP. Can Is there anything else we should handle from this PR?

**Comment by architkulkarni** [comment]
Before merging this PR, the PR https://github.com/ray-project/ray/pull/13565 needs to be merged, so that tests on the current PR work (since shutdown() is called between each test).   

After that, this PR is ready to be merged.

**Comment by rkooo567** [comment]
Is #13565 regression? Otherwise, we can just merge this PR first right?

**Review by edoakes — APPROVED** [review]

**Comment by edoakes** [comment]
Relevant tests failing @architkulkarni

**Comment by architkulkarni** [comment]
Yes, once #13565 is merged this PR will pass tests.  The relevant test currently passes in isolation but not in CI because ray.shutdown() is called in between tests.

**Comment by rkooo567** [comment]
cc@edoakes 

@architkulkarni Can you just refactor the test to not do ray.init & ray.shutdown again and again and just merge the PR first? (so that we can unblock Allen). Lingxuan also thinks the current fix can cause some issues, so it'd take some time to resolve the issue.

**Comment by architkulkarni** [comment]
Good idea.  I moved the relevant test to the beginning of the file so that it runs first, before any `ray.shutdown()` calls.

**Comment by architkulkarni** [comment]
Only `gcs_server_test` and `test_object_spilling` are failing now, and I think those are unrelated and known to be flaky.  So this should be ready to merge.

**Pull request merged**
